libbeat/publisher/queue/queue.go

```diff
@@ -80,23 +80,34 @@ type ProducerConfig struct {
 	DropOnCancel bool
 }
 
-// Producer interface to be used by the pipelines client to forward events to be
-// published to the queue.
-// When a producer calls `Cancel`, it's up to the queue to send or remove
-// events not yet ACKed.
-// Note: A queue is still allowed to send the ACK signal after Cancel. The
-//       pipeline client must filter out ACKs after cancel.
+// Producer is an interface to be used by the pipelines client to forward
+// events to a queue.
 type Producer interface {
+	// Publish adds an event to the queue, blocking if necessary, and returns
+	// true on success.
 	Publish(event publisher.Event) bool
+
+	// TryPublish adds an event to the queue, returning immediately if the queue
+	// is full, and returns true on success.
```

Comment:
more correct might be: 'if the event can not be processed by the queue'.
e.g. memqueue broker also does some cleanup. During this phase it is possible that we can't accept events, although there is some space left.

Reply:
Well, if possible I'd like somethng more precise than "can not be processed," I'm trying to articulate the specific contract for this interface. Can be we be more specific about the possible reasons for failure? Whether to call `Publish` or `TryPublish` is decided by whether the client config for the pipeline sets `PublishMode` to `DropIfFull`. If that setting is actually dropping events for reasons other than a full queue, I'd like to make it more explicit in the description.

Reply:
It depends per queue implementation. Important is to not block the caller at all if the event can not be processed right now. `DropIfFull` is maybe not a good name to transfer this message.

One use-case is packetbeat. In packetbeat we must not block if the event can not be consumed right now. Packetbeat would rather drop the event and continue processing the next network packet, then drop a a multiple network packet . This is required due to packetbeat accumulating quite some internal state and packet loss can be much more severe.

Reply:
In the abstract it depends on queue implementation, but more pragmatically we only have one non-beta implementation of `queue.Queue` in the entire codebase, so it seems desirable to be as specific as we can about why `TryPublish` might fail. (The behavior of the whole publisher pipeline depends on the contract for `Publish` and `TryPublish`, which is currently entirely implicit, making it hard to safely reason about new queue implementations.)

If the current design doesn't allow for something more specific, how about we explicitly call out that behavior dependency: `TryPublish adds an event to the queue if doing so will not block the caller, otherwise it immediately returns. The reasons a publish attempt might block are defined by the specific queue implementation and its configuration. Returns true if the event was successfully added, false otherwise.`

Reply:
SGTM